Updated API Design Guidelines (#11115)
galaxyproject/galaxy: doc/source/dev/api_guidelines.rst

the Galaxy API. The Galaxy API has grown in an ad-hoc fashion over time by
many contributors and so clients SHOULD NOT expect the API will conform to
these guidelines - but developers contributing to the Galaxy API SHOULD follow
these guidelines.

Before Getting Started
~~~~~~~~~~~~~~~~~~~~~~

Developers should familiarize themselves with HTTP method definitions 
https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9. While Galaxy
follows these imperfectly - a serious effort should be made to use HTTP
methods (GET, POST, PUT, DELETE, etc.) correctly and consistently. More
information about these methods with a focus on RESTful APIs can be found
at https://restfulapi.net/http-methods/.

Developers should familiarize themselves with the HTTP status code definitions
http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html. The API responses
should properly set the status code according to the result - in particular
2XX responses should be used for successful requests, 4XX for various
kinds of client errors, and 5XX for the errors on the server side.

Developers should also familiarize themselves with RESTful API design

> https://restfulapi.net/rest-api-design-tutorial-with-example/. Again Galaxy

Reviewer: Minor concern. Who is restfulapi.net? Should we consider using a more transparent source? This one says nothing about who the authors are, and their "about us" page, while not providing any information, claims that everything else is "...polluted with duplicate, irrelevant and poor quality information" and presents their site as the source of truth. Even though the content is good, maybe use something else as a reference in our documentation.

There are many good sources. Maybe something like [this](https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design) or [this](https://github.com/Microsoft/api-guidelines/blob/master/Guidelines.md); there's Roy Fielding's dissertation, etc. Anything, really, just not both anonymous and not peer-reviewed.

Author: You're absolutely right - I'll update that link to https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design. The resource looks more complete as well. Thanks!

Reviewer: Thank you!

follows generic RESTful guidelines imperfectly - but a serious effort should
be made to design the API as RESTfully and consistently as possible.

Documenting the API
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

API actions should include high-level docstring documentation for consumption
at docs.galaxyproject.org.

API functionality should consume and produce Pydantic_ `models
<https://pydantic-docs.helpmanual.io/usage/models/>`__ with typed and annotated
`fields <https://pydantic-docs.helpmanual.io/usage/types/>`__. This information

> will be used to generate detailed OpenAPI documentation for APIs in

Reviewer: ```suggestion
will be used to generate detailed [OpenAPI](https://www.openapis.org) documentation for APIs in
```
I think this link is relevant and may be helpful.

possible. Most of the logic should be pushed into components that
are decoupled from the actual web application and the web framework.
The role of the controller should be simply to adapt the web request
(tied to the web framework) to simple objects (plain Python objects,
Pydantic_ models, or SQL Alchemy models), send these objects to a
Python component lower in the stack, and finally to translate the
response to a web response. This design will ensure that Galaxy
application logic is usable in as many contexts as possible and
decouple the core "business logic" of Galaxy from the web framework
hosting it.

Going forward, we anticipate the most typical way this abstract
advice translates to concrete code flow is that API controllers
in :mod:`galaxy.webapps.galaxy.api` receive or produce Pydantic_
models and pass them to manager components defined in
:mod:`galaxy.managers`. The manager is then responsible for
interacting with the whole backend or specific components to
fullfil the desired request.

Galaxy work contexts provide a way for API requests (and other
components) to describe what they require from the framework

> in terms the request (examples include the Galaxy app, a user

Reviewer: ```suggestion
in terms of the request (examples include the Galaxy app, a user
```

API Design Guidelines
=====================

The following section outlines guidelines related to extending and/or modifying
the Galaxy API. The Galaxy API has grown in an ad-hoc fashion over time by

> many contributors and so clients SHOULD NOT expect the API will conform to

Reviewer 2: ```suggestion
the Galaxy API. The Galaxy API has been developed over time in an ad-hoc fashion and by
many contributors, so clients SHOULD NOT expect the API will conform to
```

these guidelines - but developers contributing to the Galaxy API SHOULD follow
these guidelines.

Before Getting Started
~~~~~~~~~~~~~~~~~~~~~~

Developers should familiarize themselves with HTTP method definitions 

> https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9. While Galaxy

Reviewer 2: ```suggestion
Developers should familiarize themselves with `HTTP method definitions
<https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9>`__. While Galaxy
```

follows these imperfectly - a serious effort should be made to use HTTP
methods (GET, POST, PUT, DELETE, etc.) correctly and consistently. More
information about these methods with a focus on RESTful APIs can be found

> at https://restfulapi.net/http-methods/.

Reviewer: ```suggestion
at https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#conform-to-http-semantics .
```
(As per discussion on resgulapi.net) + this part of Microsoft's guide is, I think, exactly "...these methods with a focus on RESTful APIs"

Author: I didn't catch that was still there - thanks for catching that.

Developers should familiarize themselves with the HTTP status code definitions

> http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html. The API responses

Reviewer 2: ```suggestion
Developers should also familiarize themselves with the `HTTP status code definitions
<http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html>`__. The API responses
```

should properly set the status code according to the result - in particular
2XX responses should be used for successful requests, 4XX for various
kinds of client errors, and 5XX for the errors on the server side.

Developers should also familiarize themselves with RESTful API design

> https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design.

Reviewer 2: ```suggestion
Finally, developers should familiarize themselves with `RESTful API design
<https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design>`__.
```

> Again Galaxy follows generic RESTful guidelines imperfectly - but a serious

Reviewer 2: ```suggestion
Again, Galaxy follows generic RESTful guidelines imperfectly, but a serious
```

effort should be made to design the API as RESTfully and consistently as possible.

Documenting the API
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> API actions should include high-level docstring documentation for consumption

Reviewer 2: ```suggestion
API endpoints should include high-level docstring documentation for consumption
```

Author: The endpoint is like an abstract thing the client communicates - I think I explicitly chose action here because I was referring to the Python method.

Reviewer 2: "An API endpoint implementation" ? API action seems a bit obscure to me.

Author: Done.

at docs.galaxyproject.org.

API functionality should consume and produce Pydantic_ `models
<https://pydantic-docs.helpmanual.io/usage/models/>`__ with typed and annotated
`fields <https://pydantic-docs.helpmanual.io/usage/types/>`__. This information
will be used to generate detailed OpenAPI_ documentation for APIs in
the near future. Typed models also improve the quality and thoroughness
of the static checking that can be done for the Galaxy backend.

To further document your API and allow other developers to see how it functions,
new API functionality should include functional tests. These functional
tests should be implemented in Python and added to either the API test suite

> or the integration test suite. Checkout the :ref:`Writing Tests for Galaxy <./writing_tests.md>`

Reviewer 2: ```suggestion
or the integration test suite. Checkout the :doc:`Writing Tests for Galaxy <writing_tests>`
```

Error responses should not consist of plain text strings - they should be
dictionaries describing the error and containing the following:

.. code-block: json

    {
      "status_code": 400,
      "err_code": 400007,
      "err_msg": "Request contained invalid parameter, action could not be completed.",
      "type": "error",
      "extra_error_info": "Extra information."
    }

The error code (``err_code``) is defined by the exception class
and pulled from ``lig/galaxy/exceptions/error_codes.json``. This JSON
document should be used to add new error conditions or improve
default error messages.

If there is an error processing some part of the request (one item in a list
for instance), the status code should be set to reflect the error and the
partial result may or may not be returned depending on the controller -
this behavior should be documented.

Backward Compatibility
~~~~~~~~~~~~~~~~~~~~~~

Backward compatibility is important and should be maintained when possible.
If changing behavior in a non-backward compatible way please ensure one
of the following holds - there is a strong reason to believe no consumers

> depend on a behavior, the behavior is effectively broken, or the API

Reviewer 2: ```suggestion
depend on the old behavior, the behavior is effectively broken, or the API
```

method being modified has not been part of a tagged dist release. If these
conditions do not hold, please help the Galaxy release manager ensure
information about the breaking API is highlighted in the Galaxy release
notes.

Whenever functionality needs to be added, removed, or modified that
breaks backward compatibility in a significant way to a component used by
the community - a "dev" variant of the API should be established and
the community should be alerted and given a time frame for when the old
behavior will be replaced with the new behavior.

Design
~~~~~~

Functionality should not be copied and pasted between controllers -
consider refactoring functionality into associated classes or short of
that into mixins_ or into managers (:mod:`galaxy.managers`).

To whatever degree is possible, the Galaxy API controller classes
(found in :mod:`galaxy.webapps.galaxy.api`) should be as lean as
possible. Most of the logic should be pushed into components that
are decoupled from the actual web application and the web framework.
The role of the controller should be simply to adapt the web request
(tied to the web framework) to simple objects (plain Python objects,

> Pydantic_ models, or SQL Alchemy models), send these objects to a

Reviewer 2: ```suggestion
Pydantic_ models, or SQLAlchemy models), send these objects to a
```
